harvester/harvester
# Add `containerd-registry` setting

**Problem:**
Users didn't have a convenient way to set RKE2 private registry.

**Solution:**
Add a setting `containerd-registry` for it.

**Related Issue:**
https://github.com/harvester/harvester/issues/2176

**Test plan:**

Setup:
* Use vagrant-pxe-harvester to create a harvester cluster.
* Create another VM `myregistry` and set it in the same virtual network.
* In `myregistry` VM:
  * Install docker.
  * Run following commands:
```
mkdir auth
 docker run \
  --entrypoint htpasswd \
  httpd:2 -Bbn testuser testpassword > auth/htpasswd

mkdir -p certs

openssl req \
  -newkey rsa:4096 -nodes -sha256 -keyout certs/domain.key \
  -addext "subjectAltName = DNS:myregistry.local" \
  -x509 -days 365 -out certs/domain.crt

sudo mkdir -p /etc/docker/certs.d/myregistry.local:5000
sudo cp certs/domain.crt /etc/docker/certs.d/myregistry.local:5000/domain.crt

docker run -d \
  -p 5000:5000 \
  --restart=always \
  --name registry \
  -v "$(pwd)"/certs:/certs \
  -v "$(pwd)"/registry:/var/lib/registry \
  -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt \
  -e REGISTRY_HTTP_TLS_KEY=/certs/domain.key \
  -v "$(pwd)"/auth:/auth \
  -e "REGISTRY_AUTH=htpasswd" \
  -e "REGISTRY_AUTH_HTPASSWD_REALM=Registry Realm" \
  -e REGISTRY_AUTH_HTPASSWD_PATH=/auth/htpasswd \
  registry:2
```
* Set IP and domain in `/etc/hosts` to all VM (harvester and `myregistry`). Remember to change the IP.
```
# vim /etc/hosts
192.168.0.50 myregistry.local
```
* Login, pull, and push nginx image in `myregsitry` VM:
```
# username: testuser, password: testpassword
docker login myregistry.local:5000
docker pull nginx:latest
docker tag nginx:latest myregistry.local:5000/nginx:latest
docker push myregistry.local:5000/nginx:latest
docker pull nginx:1.22
docker tag nginx:latest myregistry.local:5000/nginx:1.22
docker push myregistry.local:5000/nginx:1.22
```
* Copy `certs/domain.crt` content in `myregistry` VM and paste it to `additional-ca` setting.

Case 1: Change `containerd-registry` to our private registry
* Update `containerd-registry` setting as following:
```
...
value: '{"Mirrors":{"docker.io":{"Endpoints":["https://myregistry.local:5000/"]}},"Configs":{"myregistry.local:5000":{"Auth":{"Username":"testuser","Password":"testpassword"}}}}'
```
* Check content in `cattle-system/harvester-containerd-registry` secret is changed.
* Check there are new jobs to auto apply the new `containerd-registry` setting.
* Apply following yaml and nginx can be deployed.
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myregistry-nginx
spec:
  selector:
    matchLabels:
      image: myregistry-nginx
  template:
    metadata:
      labels:
        image: myregistry-nginx
    spec:
      containers:
        - imagePullPolicy: Always
          image: myregistry.local:5000/nginx:latest
          name: nginx
```

Case 2: Change `containerd-registry` to default value
* Update `containerd-registry` setting:
```
...
value: ""
```
* Check content in `cattle-system/harvester-containerd-registry` secret is changed.
* Check there are new jobs to auto apply the new `containerd-registry` setting.
* Apply following yaml and nginx 1.22 **_cannot_** be deployed.
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: myregistry-nginx-1-22
spec:
  selector:
    matchLabels:
      image: myregistry-nginx-1-22
  template:
    metadata:
      labels:
        image: myregistry-nginx-1-22
    spec:
      containers:
        - imagePullPolicy: Always
          image: myregistry.local:5000/nginx:1.22
          name: nginx
```

Case 3: Set `containerd-registry` setting back to our private registry
* Update `containerd-registry` setting as following:
```
...
value: '{"Mirrors":{"docker.io":{"Endpoints":["https://myregistry.local:5000/"]}},"Configs":{"myregistry.local:5000":{"Auth":{"Username":"testuser","Password":"testpassword"}}}}'
```
* Check content in `cattle-system/harvester-containerd-registry` secret is changed.
* Check there are new jobs to auto apply the new `containerd-registry` setting.
* Check nginx 1.22 deployment is running.
